### PR TITLE
High-resolution zoom of colorfill plots

### PR DESCRIPTION
**Pull request by quantumsteve**
**Description of work.**

The uses the [ModestImage library](https://github.com/ChrisBeaumont/mpl-modest-image/) to resample data for the screen size, replacing our similar attempt in #24205.

I modified the `ModestImage` tests to use `unittest`. The parameterized tests are run iff the `parameterized` library is present.

**To test:**

Run script in Workbench

python
```
# The following line helps with future compatibility with Python 3
# print must now be used as a function, e.g print('Hello','World')
from __future__ import (absolute_import, division, print_function, unicode_literals)

# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

LoadEventNexus(Filename="PG3_37861",OutputWorkspace="ws1")
ConvertUnits(InputWorkspace="ws1",OutputWorkspace="ws2",target="dSpacing")
Rebin(InputWorkspace="ws2", OutputWorkspace="ws3", Params='.2,0.01,2.5', PreserveEvents=False)
```

Right-click `ws3` and select `plot->colorfill`.

Use the zoom rectangle tool and verify the resulting figure maintains the resolution of the original workspace and contains the correct axes limits.

*There is no associated issue.*

*This does not require release notes* because internal change.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by peterfpeterson — CHANGES_REQUESTED** [review]
Zoom works, but pan will (intermittently) leave white space on the edges of parts of the image.

**Comment by quantumsteve** [comment]
@martyngigg I'm at a loss for fixing the tests added in #24414. Any ideas?

**Comment by martyngigg** [comment]
@quantumsteve That's odd. It looks like the upper y-limit is no longer correct. That test was supposed to check that the `update_colorplot_datalimits` functions correctly reset the axes limits based on the current dataset. As the maximum value on the Y-axis of `ws2d_histo` is `8` I thought it should be that.

It essentially calls `axes.update_datalim` with the new values. Could the new extent calculation be conflicting in some way?

**Comment by quantumsteve** [comment]
@peterfpeterson Would you please try again?

**Comment by NickDraper** [comment]
I tried this using this script (just a change of file for one available at ISIS)
``` python
Load(Filename="SANS2D00014271.nxs",OutputWorkspace="ws1")
ConvertUnits(InputWorkspace="ws1",OutputWorkspace="ws2",target="dSpacing")
Rebin(InputWorkspace="ws2", OutputWorkspace="ws3", Params='.2,0.01,2.5', PreserveEvents=False)
```

When I right click and select colour fill I get this in the log.

```
Traceback (most recent call last):
  File "c:\mantid\code\qt\applications\workbench\workbench\plugins\workspacewidget.py", line 101, in _do_plot_colorfill
    pcolormesh(self._ads.retrieveWorkspaces(names, unrollGroups=True))
  File "c:\mantid\code\qt\python\mantidqt\plotting\functions.py", line 230, in pcolormesh
    fig, axes, nrows, ncols = _create_subplots(workspaces_len, fig=fig)
  File "c:\mantid\code\qt\python\mantidqt\plotting\functions.py", line 357, in _create_subplots
    fig = plt.figure()
  File "C:\Mantid\Code\external\src\ThirdParty\lib\python2.7\lib\site-packages\matplotlib\pyplot.py", line 533, in figure
    **kwargs)
  File "c:\mantid\code\qt\applications\workbench\workbench\plotting\qappthreadcall.py", line 48, in __call__
    return self.callee(*args, **kwargs)
  File "c:\mantid\code\qt\applications\workbench\workbench\plotting\figuremanager.py", line 364, in new_figure_manager
    return new_figure_manager_given_figure(num, this_fig)
  File "c:\mantid\code\qt\applications\workbench\workbench\plotting\figuremanager.py", line 372, in new_figure_manager_given_figure
    manager = FigureManagerWorkbench(canvas, num)
  File "c:\mantid\code\qt\applications\workbench\workbench\plotting\figuremanager.py", line 185, in __init__
    self.fit_browser = FitPropertyBrowser(canvas)
TypeError: __init__() takes at least 3 arguments (2 given)

```

**Comment by martyngigg** [comment]
Should this still be in release 4.0? It's target branch will need updating if so.

File: Framework/PythonInterface/test/python/mantid/plots/modest_image/test_modest_image.py [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+import matplotlib
+matplotlib.use("agg")
+
+from matplotlib import pyplot as plt
+import matplotlib.image as mi
+
+import numpy as np
+
+from mantid.plots.modest_image import ModestImage
+run_parameterized_tests = True
```
Comment (peterfpeterson) [inline Framework/PythonInterface/test/python/mantid/plots/modest_image/test_modest_image.py]: Please add a comment here that the `parameterized` module isn't available on all the platforms we support, specific call-out to RHEL7 wouldn't be out of line.

**Review by peterfpeterson — CHANGES_REQUESTED** [review]
Overall looks good and works well for things that have constant binning. Just a minor request for a comment about the optional package.

**Review by peterfpeterson — DISMISSED** [review]

File: buildconfig/dev-packages/deb/mantid-developer/ns-control [hunk 2]
```diff
@@ -53,6 +53,7 @@ Depends: git,
   python-dateutil,
   python-sphinx-bootstrap-theme,
   python-matplotlib,
+  python-parameterized
```
Comment (rosswhitfield) [inline buildconfig/dev-packages/deb/mantid-developer/ns-control]: This package doesn't exist on Ubuntu 16.04

Comment (quantumsteve) [reply buildconfig/dev-packages/deb/mantid-developer/ns-control]: I moved the two tests using `parameterized` to for loops with no dependencies.

**Review by peterfpeterson — DISMISSED** [review]

**Comment by martyngigg** [comment]
Does this mean we can now stop shipping `skimage` with macOS and Windows and drop it from the requirements for Linuxes?

**Comment by quantumsteve** [comment]
@martyngiggYes,  ModestImage doesn't use `scikit-image` and therefore Mantid no longer depends on it. May we save that work for a separate smaller PR?

**Comment by martyngigg** [comment]
@quantumsteve Absolutely. I just wanted to check I was keeping up with things :)

File: Framework/PythonInterface/mantid/plots/modest_image/modest_image.py [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+        # return without doing anything further.
+        if (self._bounds is not None and
+                sx >= self._sx and sy >= self._sy and
+                x0 >= self._bounds[0] and x1 <= self._bounds[1] and
+                y0 >= self._bounds[2] and y1 <= self._bounds[3]):
+            return
+
+        # Slice the array using the slices determined previously to optimally
+        # match the display
+        self._A = self._full_res[y0:y1:sy, x0:x1:sx]
```
Comment (quantumsteve) [inline Framework/PythonInterface/mantid/plots/modest_image/modest_image.py]: I think `skimage.transform.resize` here could give additional options (for example the documentation suggests [using Gaussian smoothing when downsampling](http://scikit-image.org/docs/stable/auto_examples/transform/plot_rescale.html)), but I haven't looked to see if it makes 
a significant difference in our colorfill plots or measured if it significantly slows down plotting?

**Review by peterfpeterson — APPROVED** [review]
Should be good to go. Will eventually need a better solution for logorithmic binning.

**Pull request merged**